core/src/main/java/org/opensearch/sql/utils/JsonUtils.java

```diff
@@ -0,0 +1,202 @@
+      throw new SemanticCheckException(String.format(errorFormat, json, e.getMessage()), e);
+    }
+  }
+
+  private static JsonNode jsonToNode(ExprValue json) {
+    return jsonStringToNode(json.stringValue());
+  }
+
+  private static JsonNode jsonStringToNode(String jsonString) {
+    ObjectMapper objectMapper = new ObjectMapper();
+    JsonNode jsonNode;
+    try {
+      jsonNode = objectMapper.readTree(jsonString);
+    } catch (JsonProcessingException e) {
+      final String errorFormat = "JSON string '%s' is not valid. Error details: %s";
+      throw new SemanticCheckException(String.format(errorFormat, jsonString, e.getMessage()), e);
+    }
+    return jsonNode;
+  }
+
+  private static ExprValue processJsonNode(JsonNode jsonNode) {
+    if (jsonNode.isFloatingPointNumber()) {
+      return new ExprDoubleValue(jsonNode.asDouble());
+    }
+    if (jsonNode.isIntegralNumber()) {
+      return new ExprIntegerValue(jsonNode.asLong());
+    }
+    if (jsonNode.isBoolean()) {
+      return jsonNode.asBoolean() ? LITERAL_TRUE : LITERAL_FALSE;
+    }
+    if (jsonNode.isTextual()) {
+      return new ExprStringValue(jsonNode.asText());
+    }
+    if (jsonNode.isArray()) {
+      List<ExprValue> elements = new LinkedList<>();
+      for (var iter = jsonNode.iterator(); iter.hasNext(); ) {
+        jsonNode = iter.next();
+        elements.add(processJsonNode(jsonNode));
+      }
+      return new ExprCollectionValue(elements);
+    }
+    if (jsonNode.isObject()) {
+      Map<String, ExprValue> values = new LinkedHashMap<>();
+      for (var iter = jsonNode.fields(); iter.hasNext(); ) {
+        Map.Entry<String, JsonNode> entry = iter.next();
+        values.put(entry.getKey(), processJsonNode(entry.getValue()));
+      }
+      return ExprTupleValue.fromExprValueMap(values);
+    }
+
+    // in all other cases, return null
+    return LITERAL_NULL;
+  }
+
+  /** Converts a JSON encoded string to an Expression object. */
+  public static ExprValue setJson(ExprValue json, ExprValue path, ExprValue valueToInsert) {
+
+    String jsonUnquoted = json.stringValue();
+    String pathUnquoted = path.stringValue();
+    Object valueUnquoted = valueToInsert.value();
+    Configuration conf =
+        Configuration.defaultConfiguration().addOptions(Option.SUPPRESS_EXCEPTIONS);
+    try {
+      JsonPath jsonPath = JsonPath.compile(pathUnquoted);
+      DocumentContext docContext = JsonPath.using(conf).parse(jsonUnquoted);
+      Object readResult = docContext.read(jsonPath);
+      if (readResult == null) {
+        recursiveCreate(docContext, pathUnquoted, valueUnquoted).jsonString();
+      } else if (readResult instanceof String) {
+        docContext.set(pathUnquoted, valueUnquoted).jsonString();
+      } else {
+        if (((List<?>) readResult).isEmpty()) {
+          docContext.add(pathUnquoted, valueUnquoted).jsonString();
+        } else {
+          docContext.set(pathUnquoted, valueUnquoted).jsonString();
+        }
+      }
```

Comment:
Can reduce nesting here?

```suggestion
      if (readResult == null) {
        recursiveCreate(docContext, pathUnquoted, valueUnquoted).jsonString();
      } else if (readResult instanceof String) {
        docContext.set(pathUnquoted, valueUnquoted).jsonString();
      } else if (((List<?>) readResult).isEmpty()) {
        docContext.add(pathUnquoted, valueUnquoted).jsonString();
      } else {
        docContext.set(pathUnquoted, valueUnquoted).jsonString();
      }
```

Reply:
Addressed. Thanks. Please resolve.